public/app/plugins/datasource/loki/formatterTS.ts

```diff
@@ -0,0 +1,406 @@
+  const hasString = node.getChild(String);
+
+  if (hasString) {
+    const parserNode = (node.getChild(Regexp) || node.getChild(Pattern))!;
+    const stringNode = node.getChild(String)!;
+
+    const parser = query.substring(parserNode.from, parserNode.to);
+    const string = query.substring(stringNode.from, stringNode.to);
+
+    return `| ${parser}${string}`;
+  }
+
+  const labelParser = query.substring(node.from, node.to);
+  return `| ${labelParser}`;
+}
+
+export function formatJsonExpressionParser(node: SyntaxNode, query: string): string {
+  const jsonExpressionNodes = iterateNode(node, [JsonExpression]);
+  let response = '';
+
+  jsonExpressionNodes.forEach((node) => {
+    const identifierNode = node.getChild(Identifier)!;
+    const valueNode = node.getChild(String)!;
+
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    response += `${identifier}=${value}, `;
+  });
+
+  return `| json ${trimEnd(response, ', ')}`;
+}
+
+export function formatLabelFilter(node: SyntaxNode, query: string): string {
+  if (node.getChild(Matcher)) {
+    const matcherNode = node.getChild(Matcher)!;
+
+    const identifierNode = matcherNode.getChild(Identifier)!;
+    const valueNode = matcherNode.getChild(String)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (matcherNode.getChild(Eq)) {
+      return `| ${identifier}=${value}`;
+    } else if (matcherNode.getChild(Neq)) {
+      return `| ${identifier}!=${value}`;
+    } else if (matcherNode.getChild(Re)) {
+      return `| ${identifier}=~${value}`;
+    } else if (matcherNode.getChild(Nre)) {
+      return `| ${identifier}!~${value}`;
+    }
+  }
+
+  if (node.getChild(IpLabelFilter)) {
+    const ipLabelFilterNode = node.getChild(IpLabelFilter)!;
+
+    const identifierNode = ipLabelFilterNode.getChild(Identifier)!;
+    const valueNode = ipLabelFilterNode.getChild(String)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (ipLabelFilterNode.getChild(Eq)) {
+      return `| ${identifier}=ip(${value})`;
+    } else if (ipLabelFilterNode.getChild(Neq)) {
+      return `| ${identifier}!=ip(${value})`;
+    }
+  }
+
+  if (node.getChild(UnitFilter) && node.getChild(UnitFilter)!.getChild(DurationFilter)) {
+    const subNode = node.getChild(UnitFilter)!;
+    const durationFilterNode = subNode.getChild(DurationFilter)!;
+
+    const identifierNode = durationFilterNode.getChild(Identifier)!;
+    const valueNode = durationFilterNode.getChild(Duration)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (durationFilterNode.getChild(Gtr)) {
```

Comment:
Here you could also use something like `nextSibling` and avoid the literal check.
(Apologies for the repeated comments, but don't want you to miss this).